RFC Local import (#5)
tweag/ghc-proposals: proposals/0000-local-import.rst

            **After creating the pull request, edit this file again, update the
            number in the link, and delete this bold sentence.**
.. sectnum::
.. contents::

We propose to allow the use of ``import`` statements in any ``let``, ``where`` or ``do`` block. The effect of those imports is limited to the corresponding block and (for ``let`` and ``where`` blocks) to the defined expression. Moreover, we define a shortcut syntax, ``Qualifier.{ … }``, to allow scoping a local import over a single expression.

For example, the following snippets are considered valid:

::

  module MyModule where

  import Data.Set (Set)
  import qualified Data.Set as Set
  
  foo :: (Set Int, Set String)
  foo = (s1, fromList ["foo", "bar", "baz"])
    where
      import Set
      s1 = fromList [1, 2, 3]

::

  module Main where
  import System.Environment as Env
   
  main :: IO ()
  main = do
    args <- Env.getArgs
    case args of
      x:y:_ -> …
      _ -> do
        import System.Exit as Exit
        Exit.{ exitWith $ ExitFailure (-1) }

Motivation
------------

The motivations are the following:

- Making import lists smaller and easier to maintain.

> - Making it easier to understand why a module is imported.

Reviewer: ... and when the imported symbols are used. 

(Or something similar)

- Module qualifiers can severely affect readability.

The latter problem is often addressed by choosing one-letter qualifiers, which can only be understood by looking up their definition at the top of the file. It is an imperfect solution at best, and does not work well for operators ; for this reason, operators are frequently overloaded via lawless typeclasses for mere syntactic convenience.

OCaml 3.12, released in 2011, introduced two syntactic contructs for locally opening modules: ``let open Module in …`` and ``Module.( … )``. In less than 10 years, this feature has gained widespread adoption throughout the ecosystem, bringing significant benefits in terms of readability and maintainability. For the reasons stated above, we believe Haskell should follow its example.

Proposed Change Specification
-----------------------------
We allow qualified and non-qualified ``import`` statements at the beginning of any ``let``/``where``/``do``-block.

For all constructs, the effect of such imports is limited to the scope the enclosing block. For ``let`` and ``where``, the imports also affect the corresponding expression. For example:
::

>   let import Foo in <expr>

Reviewer: The examples implies that `let` and `where` can contain `import` statement and other statements. However there is no discussion about the ordering of them. Is the following allowed:

```
let
    import A
    foo = bar
    import B
```

Or should `import` appear only in first position of the `let` / `where` block?

Same question for the `do` blocks. Does it affect all the block, or just the following statements? Can they appear everywhere in a `do` block or just at the beginning?

Author: I believe those should only be allowed at the beginning of a `let` or `where` block.

1) This is consistent with Haskell modules as a whole. I want to be as consistent as possible with existing restrictions.

2) Declarations in `let` and `where` clauses can be mutually recursive and appear in any order ; OTOH, in:

```
let
    import A
    foo = bar
    import B
```
it would feel quite unnatural to allow `bar` to use symbols from `B`.

Author: I have put emphasis on *beginning*.

::

  f = <expr>
    where import Foo

Moreover, ``import`` statements are allowed to refer to any module qualifier specified outside that block (that is, local ``import`` statements are not limited to full module names). This means that the following is allowed:
::

  import qualified Data.Set as Set
  foo = …
    where import Set
          …
As well as:
::

  foo = …
    where import qualified Data.Set as Set
          bar = …
            where import Set
                  …
But the following is not valid:
::

  foo = …
    where import qualified Data.Set as Set
          import Set
This is consistent with today's semantics for ``import``, which does not allow for:
::

  module MyModule where

  import qualified Data.Set as Set
  import Set

Finally, we allow the syntactic shortcut ``Qualifier.{ <expression> }``, which simply desugars to:
::

  let import Qualifier in <expression>

The following changes in the Haskell 2010 grammar are required:

::

  decls → { impdecls ; decl_1 ; … ; decl_n }    (n ≥ 0)
       | { decl_1 ; … ; decl_n }
  lexp → …
       | modid.{ lexp }
       | do { stmts }
       | do { impdecls ; stmts }

Effect and Interactions
-----------------------

> Detail how the proposed change addresses the original problem raised in the motivation.

Reviewer: Put emphasis here on the fact that it does not change the current semantic, just add more feature.

Local import
==============

.. proposal-number:: Leave blank. This will be filled in when the proposal is
                     accepted.
.. trac-ticket:: Leave blank. This will eventually be filled with the Trac
                 ticket number which will track the progress of the
                 implementation of the feature.
.. implemented:: Leave blank. This will be filled in with the first GHC version which
                 implements the described feature.
.. highlight:: haskell
.. header:: This proposal is `discussed at this pull request <https://github.com/ghc-proposals/ghc-proposals/pull/0>`_.
            **After creating the pull request, edit this file again, update the
            number in the link, and delete this bold sentence.**
.. sectnum::
.. contents::

> We propose to allow the use of ``import`` statements in any ``let``, ``where`` or ``do`` block. The effect of those imports is limited to the corresponding block and (for ``let`` and ``where`` blocks) to the defined expression. Moreover, we define a shortcut syntax, ``Qualifier.{ … }``, to allow scoping a local import over a single expression.

Reviewer 2: To avoid an inevitable question it might be worth mentioning that `Qualified.( … )` can't be used because it's already valid haskell and with a different meaning (I assume that it's the reason for the choice of the braces)

Author: :+1:

For example, the following snippets are considered valid:

::

  module MyModule where

  import Data.Set (Set)
  import qualified Data.Set as Set
  
  foo :: (Set Int, Set String)
  foo = (s1, fromList ["foo", "bar", "baz"])
    where
      import Set
      s1 = fromList [1, 2, 3]

::

  module Main where
  import System.Environment as Env
   
  main :: IO ()
  main = do
    args <- Env.getArgs
    case args of
      x:y:_ -> …
      _ -> do
        import System.Exit as Exit

>         Exit.{ exitWith $ ExitFailure (-1) }

Reviewer 2: This example is slightly contrived since the last `Exit.{ }` is totally superfluous (because you didn't import `System.Exit` qualified so it's already in the scope)

- Explicit import lists are inconvenient (they must be edited whenever the code changes). 
- Module qualifiers can severely affect readability.

The latter problem is often addressed by choosing one-letter qualifiers, which can only be understood by looking up their definition at the top of the file. It is an imperfect solution at best, and does not work well for operators ; for this reason, operators are frequently overloaded via lawless typeclasses for mere syntactic convenience.

OCaml 3.12, released in 2011, introduced two syntactic contructs for locally opening modules: ``let open Module in …`` and ``Module.( … )``. In less than 10 years, this feature has gained widespread adoption throughout the ecosystem, bringing significant benefits in terms of readability and maintainability. For the reasons stated above, we believe Haskell should follow its example.

Proposed Change Specification
-----------------------------
We allow qualified and non-qualified ``import`` statements at the **beginning** of any ``let``/``where``/``do``-block.

> For all constructs, the effect of such imports is limited to the scope the enclosing block. For ``let`` and ``where``, the imports also affect the corresponding expression. For example:

Reviewer 2: ```suggestion
For all constructs, the effect of such imports is limited to the scope of the enclosing block. For ``let`` and ``where``, the imports also affect the corresponding expression. For example:
```

                  …
But the following is not valid:
::

  foo = …
    where import qualified Data.Set as Set
          import Set
This is consistent with today's semantics for ``import``, which does not allow for:
::

  module MyModule where

  import qualified Data.Set as Set
  import Set

Finally, we allow the syntactic shortcut ``Qualifier.{ <expression> }``, which simply desugars to:
::

  let import Qualifier in <expression>

The following changes in the Haskell 2010 grammar are required:

::

  decls → { impdecls ; decl_1 ; … ; decl_n }    (n ≥ 0)
       | { decl_1 ; … ; decl_n }
  lexp → …
       | modid.{ lexp }
       | do { stmts }
       | do { impdecls ; stmts }

Effect and Interactions
-----------------------
This proposal strictly extends the language, without affecting the behavior of existing code. It does not interact with any existing language extension.

The changes give programmers various ways to reduce the number of toplevel imports, to limit their effect to specifics parts of the code and to convey intent about their uses. The shortcut syntax can be especially useful for scoping module imports over expressions with operators in DSLs.

> Costs and Drawbacks

Reviewer 2: Another (maybe minimal) cost is that this makes the detection of dependencies harder (because we can't just look at the header anymore). This quite likely breaks some tooling which assumed that the only `import` statements where at toplevel (`snack` at least [does this](https://github.com/nmattia/snack/blob/08861356442d867d3bdcd217c4521c53264388c0/snack-lib/Imports.hs#L91-L95))

Author: This is a very good point.

-------------------
TODO: Development and maintenance costs.

We expect these changes to be easy to grasp by beginners. Moreover, they could greatly improve learnability of Haskell libraries, by expliciting the module each function comes from in examples and tutorials.

Some library writers might choose to design their library around this extension, making its use virtually unavoidable for downstream users, which could be perceived as a drawback by those unwilling to enable it.

Alternatives
------------
To our knowledge, there is no other language feature or extension providing similar benefits.

Unresolved Questions
--------------------
It might be valuable to allow some form of typelevel local import ; the shortcut syntax, in particular, could be used to simplify type signatures.

> Local imports could be used to unambiguously hide globally-defined symbols. As an example, the ``blaze-html`` library provides symbols for ``head``, ``div`` and ``id`` ; for this reason,  the relevant modules are frequently imported qualified, or those symbols are explicitly hidden with ``-XNoImplicitPrelude`` and an explicit import. This is necessary from preventing uses of those symbols to be reported as ambiguous by the compiler. Without type-driven disambiguation, this is the only sane behavior in current Haskell, which only allows a single, unordered list of module imports ; however, local imports could be seen as defining nested scopes, such that:

Reviewer 2: A related questions is about the semantics of

```haskell
import Foo as A

main :: IO ()
main = do
  import Bar as A
  A.x
```

Should the `import Bar as A` shadow the outer `import Foo as A` ? Or should both `Foo` and `Bar` be available as `A` as if both imports were at toplevel? And do we want to raise a warning if `x` is defined both in `Foo` and `Bar`?

Author: > Should the import Bar as A shadow the outer import Foo as A ? Or should both Foo and Bar be available as A as if both imports were at toplevel?

The latter, in my view.

> And do we want to raise a warning if x is defined both in Foo and Bar?

Such is not the case with toplevel modules, so I don't see a reason for this to change with local imports ; however, you would get an “Ambiguous occurrence” error if you try to use that symbol.

Author: I just realized that it was a bit more subtle. Indeed, we could considered scopes as nested and allow _symbols_ from the innermost import to shadow symbols from the uppermost one (I still don't think the import should be shadowed as a whole, though).

::

  f = <expr>
    where import Foo
The set of visible typeclass instances at any given point is the union of instances defined in modules imported in all enclosing scopes.  

Moreover, ``import`` statements are allowed to refer to any module qualifier specified outside that block (that is, local ``import`` statements are not limited to full module names). This means that the following is allowed:
::

  import qualified Data.Set as Set
  foo = …
    where import Set
          …
As well as:
::

  foo = …
    where import qualified Data.Set as Set
          bar = …
            where import Set
                  …
But the following is not valid:
::

  foo = …
    where import qualified Data.Set as Set
          import Set
This is consistent with today's semantics for ``import``, which does not allow for:
::

  module MyModule where

  import qualified Data.Set as Set
  import Set

As is already the case, module qualifiers can be re-used. For example, in:
::

  import Foo as A

  main :: IO ()
  main = do
    import Bar as A
    A.x

The symbol ``x`` in ``A.x`` is searched in both ``Bar`` and ``Foo``.

> Finally, we allow the syntactic shortcut ``Qualifier.{ <expression> }``, which simply desugars to:

Reviewer 3: I think that, while convenient, I'd be slightly opposed to this aspect of the proposal since it introduces brand new syntax. I wonder about the plausibility of using e.g. quasiquoters here? Local imports would increase the utility of this ticket being implemented: https://gitlab.haskell.org/ghc/ghc/issues/1475

Author: QuasiQuoters are definitely a good idea. I have mentioned the idea in the alternatives. Thanks !

            **After creating the pull request, edit this file again, update the
            number in the link, and delete this bold sentence.**
.. sectnum::
.. contents::

This proposal allows the use of ``import`` statements in any ``let``, ``where`` or ``do`` block. The effect of those imports is limited to the corresponding block and (for ``let`` and ``where`` blocks) to the defined expression. Moreover, it defines a shortcut syntax, ``Qualifier.{ … }``, to allow scoping a local import over a single expression.

For example, the following snippets are considered valid:

::

  module MyModule where

  import Data.Set (Set)
  import qualified Data.Set as Set
  
  foo :: (Set Int, Set String)
  foo = (s1, fromList ["foo", "bar", "baz"])
    where
      import Set
      s1 = fromList [1, 2, 3]

::

  module Main where
  import System.Environment as Env
   
  main :: IO ()
  main = do
    args <- Env.getArgs
    case args of
      x:y:_ -> …
      _ -> do
        import qualified System.Exit as Exit
        Exit.{ exitWith $ ExitFailure (-1) }

Motivation
------------

> This proposal is strongly inspired by two OCaml constructs, introduced in 2011 with version 3.12: ``let open Module in …`` and ``Module.( … )``. In less than 10 years, this feature has gained widespread adoption throughout the ecosystem, bringing significant benefits in terms of readability and maintainability.

Reviewer 4: Couple of nitpicks here:

- This is not the right place for this. You can put it at the end of the summary (that's what I'd do) or at the end of the motivation section. But at any rate, it is a reference, not a starting point.
- I'd say something like « strongly inspired by Ocaml's “local open” feature which introduces two constructs[…] ».

> The proposed changes introduce Haskell equivalents in order to:

Reviewer 4: Of course, it means changing this sentence a little too.

- Make import lists smaller and easier to maintain.
- Make it easier to understand why a module is imported, and where its symbols are used.
- Make it easier to refactor a definition, along with its imports.
- Make it more practical to use qualified imports, especially with operators.

> Import lists in Haskell can quickly become unwieldly. The recent proliferation of GHC proposals intending to simplify module imports is good evidence that this feeling is widely shared: 1_, 2_, 3_.

Reviewer 4: ```suggestion
Import lists in Haskell can become unwieldly. The recent proliferation of GHC proposals intending to improve module imports is good evidence that this feeling is widely shared: 1_, 2_, 3_.
```

> Oftentimes, a module is only imported for a single symbol or two. The purpose of such imports is not obvious at first glance and can significantly clutter the import list. Moving those imports closer to the use-site would solve both of these problems by making the intent clearer and removing them from the toplevel list. Moreover, definitions would be easier to refactor, as they would be more self-contained ; for example, it would be easy to remove a definition along with its specific imports.

Reviewer 4: ```suggestion
Oftentimes, a module is only imported for a single symbol or two. The purpose of such imports is not obvious at first glance and can significantly clutter the import list. Moving those imports closer to the use-site would solve both of these problems by making the intent clearer and removing them from the toplevel list. Moreover, definitions would be easier to refactor, as they would be more self-contained; for example, it would be easy to remove a definition along with its specific imports.
```

To avoid polluting the namespace with conflicting identifiers, Haskell programmers can choose between explicit and qualified imports ; however, they are often reluctant to do so as:

- Explicit import lists are inconvenient (they must be frequently edited). 

> - Module qualifiers can severely affect readability.

Reviewer 4: I have no idea what you are saying in this paragraph+bullet list.

Moreover, ``import`` statements are allowed to refer to any module qualifier specified outside that block (that is, local ``import`` statements are not limited to full module names). This means that the following is allowed:
::

  import qualified Data.Set as Set
  foo = …
    where import Set
          …
As well as:
::

  foo = …
    where import qualified Data.Set as Set
          bar = …
            where import Set
                  …
But the following is not valid:
::

  foo = …
    where import qualified Data.Set as Set
          import Set
This is consistent with today's semantics for ``import``, which does not allow for:
::

  module MyModule where

  import qualified Data.Set as Set
  import Set

As is already the case, module qualifiers can be re-used. For example, in:
::

  import Foo as A

  main :: IO ()
  main = do
    import Bar as A
    A.x

The symbol ``x`` in ``A.x`` is searched in both ``Bar`` and ``Foo``.

Finally, the syntactic shortcut ``Qualifier.{ <expression> }`` is introduced, which simply desugars to:
::

  let import Qualifier in <expression>

> The OCaml syntax, ``Qualifier.( … )``, was not chosen as it would steal syntax (this is currently parsed as a value constructor composed with the enclosed expression).

Reviewer 4: ```suggestion
The  syntax, ``Qualifier.( … )``, user in Ocaml, was not chosen as it would steal syntax (this is currently parsed as a value constructor composed with the enclosed expression).
```

Also, move this to alternatives.

   
The following changes in the Haskell 2010 grammar are required:

::

  decls → { impdecls ; decl_1 ; … ; decl_n }    (n ≥ 0)
        | { decl_1 ; … ; decl_n }
  lexp  → …
        | modid.{ lexp }
        | do { stmts }
        | do { impdecls ; stmts }

Effect and Interactions
-----------------------
This proposal strictly extends the language, without affecting the behavior of existing code. It does not interact with any existing language extension.

The changes give programmers various ways to reduce the number of toplevel imports, to limit their effect to specifics parts of the code and to convey intent about their uses. The shortcut syntax can be especially useful for scoping module imports over expressions with operators in DSLs.

> Furthermore, it encourages the use of qualified imports, as those can be locally “de-qualified” in order to improve readability.

Reviewer 4: ```suggestion
Furthermore, it encourages the use of qualified imports, as those can be locally “unqualified” in order to improve readability.
```

Costs and Drawbacks
-------------------
These changes should be easy to grasp by beginners. Moreover, they could greatly improve learnability of Haskell libraries, by expliciting the module each function comes from in examples and tutorials.

Some existing tooling (e.g., ``snack``) assume that imports are only found at the toplevel and might be broken by this change. Perhaps more importantly, this change would make it harder for IDE-like tools such as ``hie`` to determine the set of valid completions ; such tools would need to be made context-sensitive, like OCaml's merlin.

This change would make it harder to determine at first glance inter-dependencies between modules.

> Finally, some library writers might choose to design their library around this extension. Using such libraries without this extension enabled might be inconvenient, which could be perceived as a drawback by some users.

Reviewer 4: This sounds really vague. Plus, I'd say that `vector` and `containers` (the two most popular Haskell libraries if any are) are already designed around this extension, even if it doesn't exist. And nobody is complaining.

This claim is hard to back. Let's drop it.

Reviewer 5: Lots of people are complaining, that's the point of this proposal I'd say.

> Alternatives

Reviewer 4: Split the Alternatives section with subsections.

------------

> To our knowledge, there is no other language feature or extension providing similar benefits.

Reviewer 4: Useless. Drop.

> It might be valuable to also allow some form of typelevel local import ; the shortcut syntax, in particular, could be used to simplify type signatures.

Reviewer 4: I have no idea what this means.

> The syntactic shortcut syntax is orthogonal to the rest of the proposal and could be entirely removed. The same functionality could potentially be achieved with QuasiQuoters.

Reviewer 4: I have _some_ idea what this means, but it isn't clear.

- I'm guessing « the syntactic shortcut syntax » means « the `Module.{…}` syntax ». Say it like this (or something equivalent, anyway: no vague pointers).
- But do say that in Ocaml, it has prove very useful. Give examples of why.
- I don't understand what you mean about quasiquoters. But anyway: what would it look like? what would be the drawbacks?

> As an extension to the proposed behavior, local imports could be used to shadow globally-defined symbols. As an example, the ``blaze-html`` library provides symbols for ``head``, ``div`` and ``id`` ; for this reason,  the relevant modules are frequently imported qualified, or those symbols are explicitly hidden with ``-XNoImplicitPrelude`` and an explicit import. This is necessary for preventing uses of those symbols to be reported as ambiguous by the compiler. Without type-driven disambiguation, this is the only sane behavior in current Haskell, which only allows a single, unordered list of module imports ; however, local imports could be seen as defining nested scopes, such that:

Reviewer 4: This also raises another question actually.

What does the following do:

```haskell
f = id

g = id
  where
    import Prelude hiding (id)
```

do?

(I'd imagine it doesn't hide `id` any further, it just doesn't import it a second time, but it should be specified more precisely in the specification section).

::

  {-# LANGUAGE OverloadedStrings #-}
  import Text.Blaze.Html4.Strict as Blaze
  import Text.Blaze.Html4.Strict.Attributes as Blaze

  markup :: Html
  markup = head $ div ! id "foo"
    where import Blaze
compiles without error. Similarly, in the following example:
::

  import Foo as A

  main :: IO ()
  main = do
    import Bar as A
    A.x

> If ``x`` is defined in both ``Foo`` and ``Bar``, the import from ``Bar`` could take precedence over the one from ``Foo``. Finally, DSLs could benefit from this change to override arithmetic operators without implementing bogus ``Num`` instances.

Reviewer 4: « bogus » is a loaded term, you probably want to avoid it.

Also, readers don't necessarily share where you're coming from here. Explain what such an instance looks like (with some undefined symbols). And how your idea of hierarchical shadowing could solve this problem.

We expect these changes to be easy to grasp by beginners. Moreover, they could greatly improve learnability of Haskell libraries, by expliciting the module each function comes from in examples and tutorials.

Some existing tooling (e.g., ``snack``) assume that imports are only found at the toplevel and might be broken by this change. Perhaps more importantly, this change would make it harder for IDE-like tools such as ``hie`` to determine the set of valid completions ; such tools would need to be made context-sensitive, like OCaml's merlin.

This change would make it harder to determine at first glance inter-dependencies between modules.

> Finally, some library writers might choose to design their library around this extension, making its use virtually unavoidable for downstream users, which could be perceived as a drawback by those unwilling to enable it.

Reviewer: I can imagine library design which leverage this extension, but I cannot imagine anything impossible to do without this extension. Care to elaborate?

Author: Not _impossible_ ; rather _inconvenient_. I'll make that statement clearer.
